src/Domain/GridCoordinates.hpp

```diff
@@ -0,0 +1,111 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines function GridCoordinates
+
+#pragma once
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/IndexIterator.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/CoordinateMaps/CoordinateMap.hpp"
+#include "Numerical/Spectral/LegendreGaussLobatto.hpp"
+
+template <size_t Dim1>
+class Index;
+template <size_t VolumeDim>
+class Direction;
+
+/*!
+ * \ingroup ComputationalDomain
+ * \brief Compute the grid coordinates in an Element
+ *
+ * \returns grid-frame vector holding coordinates
+ *
+ * \remarks assumes that the Mesh lies at the Legendre-Gauss-Lobatto points of
+ *  the reference cell
+ *
+ * \details
+ * Computes the grid coordinates by applying the coordinate_map at the
+ * Legendre-Gauss-Lobatto points of the reference cell.
+ *
+ * \example
+ * \snippet Test_GridCoordinates.cpp grid_coordinates_example
+ */
+template <size_t VolumeDim, typename... Maps>
+tnsr::I<DataVector, VolumeDim, Frame::Grid> grid_coordinates(
+    const Index<VolumeDim>& mesh,
+    const CoordinateMap<Frame::Logical, Frame::Grid, Maps...>& coordinate_map) {
+  tnsr::I<DataVector, VolumeDim, Frame::Logical> logical_x{};
+  tnsr::I<DataVector, VolumeDim, Frame::Grid> the_grid_coordinates{
+      DataVector(mesh.product())};
+
+  for (size_t d = 0; d < VolumeDim; ++d) {
+    logical_x.get(d) = Basis::lgl::collocation_points(mesh[d]);
+  }
+
+  for (IndexIterator<VolumeDim> index(mesh); index; ++index) {
```

Comment:
Calling the map separately for every point should not be necessary anymore.

Reply:
Yes this can be done better. I had tried to do this and was confused when it wouldn't work, and just realised it was because what's called logical_x above is not the logical coordinates, but is just a tensor of the collocation points in each dim (each DataVector within it is of a different size). The more elegant solution is then to construct the true logical coordinates, which must be done pointwise, and then call the coordinate map on the result.